### ISSUE TITLE
EX100 basic support intersted to anyone?

### ISSUE THREAD

**Issue by doctor64** [pwr-Solaar/Solaar#347]
Anybody interested in support for EX100? Its rather old Logitech mouse/keyboard combo kit. 
http://support.logitech.com/en_us/product/cordless-desktop-ex-100-product
It is not Unifying, looking to receiver I'm think it's even operated in different RF frequency range, but, fortunately it support some limited subset of HID++ 1.0 commands. At least battery status reported correctly. 
Unfortunately, adding support for EX100 in Solaar require lots of hacks and workarounds in solaar code. I'm not very happy with code quality, but if it needed to someone except me i can try to create PR.

**Comment by Lekensteyn** [comment]
What features would be interesting/worthwhile to add? Is the battery status accessible via the normal HID++ 1.0 register readouts?

**Comment by doctor64** [comment]
@Lekensteyn Yes, battery state reported and look's like accurate on both keyboard and mouse.
Also seems like keyboard have Fn Lock/Fn swap feature, but it working ... strange. Some buttons works, some not, some works but have different actions, not matching symbols on keys. So, currently only battery status implemented. I can provide pull request if you wish to look at it.

**Comment by Lekensteyn** [comment]
Device is almost 10 years old (manual mentions 2008), I am not sure if is really helpful to add. But if there is at least one (you?) who would benefit from it, why not? (Be sure to include device information if possible, like scan-registers output?)

If it is too much work for you to prepare a PR, don't bother, I am also a bit light on time ;)

**Comment by doctor64** [comment]
@Lekensteyn I will provide PR for EX100 after #337 merged, to minimize future merging efforts

**Comment by doctor64** [comment]
See PR #378

**Comment by pfps** [comment]
Is PR #378 still relevant?

**Comment by doctor64** [comment]
I have version which suit my needs. Because lack of interest and fact what my changes make solaar source more complex and difficult to understand, I'm not sure what marging EX100 support to main branch really needed. If it need to someone, i can try to prepare more polished PR

**Comment by pfps** [comment]
I may be able to take a look at that PR.  Some of my changes to make receiver c534 work better might make the changes for this receiver easier.

**Comment by pfps** [comment]
@doctor64 The recent changes to add more information about receivers in descriptors.py and better support for nano receivers like the c534 should make it easier to support the ex100.  You might want to redo your patch - I will then review it.

**Comment by doctor64** [comment]
@pfps Thank you for tip. I will try to do something.

**Comment by akumria** [comment]
FYI - I recently found one of these lying around and decided to hook it up.

I would be nice if you could at least publish what you have so-far; I rebased #378 manually and nothing actually worked.

So either I did it wrong, or something else isn't working right.

**Comment by doctor64** [comment]
@akumria Ohh, I promise I work on this starting from tomorrow :) And yes, I don't expect what direct rebasing will work, internal operation of Solaar changed in many things since my PR. Just a note, don't expect much, for me only battery reporting worked for this old stuff.

**Comment by ruifoto** [comment]
Hi, i am also interested in the support for EX100 Keyboard, since I own one. Please share any progress if made!

**Comment by doctor64** [comment]
Ok, it take some time, but its works, at least battery level reporting.

**Comment by pfps** [comment]
@ruifoto Please test out PR #859 

Try both `bin/solaar` and `bin/solaar show`.  Check that your devices show up and that all information presented looks reasonable.


To run PR #859, first clone Solaar if you have not already done so and cd to the clone directory.  The first time you download the pull request, fetch it into a new branch and checkout that branch, as in:
```
git fetch origin pull/859/head:pull_859
git checkout pull_859
```
To download a new version of the pull request, pull it into your branch for the request, as in:
```
git checkout pull_859
git pull origin pull/859/head:pull_859
```

**Comment by doctor64** [comment]
Just in case - `solaar show` output from my EX100 kit
<details>
<summary>solaar show output</summary>

```
EX100 Receiver 27 Mhz
  Device path  : /dev/hidraw3
  USB id       : 046d:c517
  Serial       : None
  Has 2 paired device(s) out of a maximum of 4.
  Notifications: wireless (0x000100)

  1: Wireless Mouse EX100
     Codename     : EX100m
     Kind         : mouse
     Wireless PID : 3F00
     Protocol     : HID++ 1.0
     Serial number: ?
     The power switch is located on the (unknown).
     Notifications: roller V, mouse extra buttons, battery status, roller H (0x3C0000).
     Features: special buttons, scroll accel (0x420000)
     Battery: full, discharging.

  3: Wireless Keyboard EX100
     Codename     : EX100
     Kind         : keyboard
     Wireless PID : 6500
     Protocol     : HID++ 1.0
     Serial number: ?
     The power switch is located on the (unknown).
     Notifications: battery status (0x100000).
     Features: special buttons, scroll accel (0x420000)
     Battery: good, discharging.
```

</details>

**Comment by ruifoto** [comment]
@pfps Thanks for the instructions. I did as you you instructed. However my EX100 Keyboard is not detected. I constantly getting the message: Pairing failed: device timeout. I did ensure the batteries are new.
The receiver that I have is the unifying receiver with the USB-ID: 046d:c52b.
@doctor64 How did you connect your EX100 Keyboard? Did you encounter any problem? Thanks.

**Comment by pfps** [comment]
@ruifoto Logitech has a habit of updating the innards of devices while keeping the model number the same.   I suspect that you have a newer version of the EX100, particularly as you are using a unifying receiver.

What does Solaar do with your EX100 without this PR?

**Comment by ruifoto** [comment]
@pfps Solaar says the same without the PR: Pairing failed: device timeout.
I had the unifying receiver because I lost my original receiver for the EX100. I thought the unifying receiver from Logitech would work out of the box. But I was wrong.
Here is a screenshoot of my EX100. I think it is the old model.
![IMG_20200719_170858](https://user-images.githubusercontent.com/10601113/87878228-3c05e580-c9e3-11ea-9f96-9f157063a727.jpg)

**Comment by doctor64** [comment]
@ruifoto No, EX100 is not compatible with unifying, sorry. It even use completely different frequencyes, 27 MHz for EX100 and 2.4GHz for Unifying. You need correct receiver, like this:
![receiver](https://secure.logitech.com/assets/33893/10/ex100-pid1.jpg)

**Comment by doctor64** [comment]
@ruifoto My keyboard marking is  M/N: Y-RGH94 P/N: 920-001224 PID: FX8461C

**Comment by doctor64** [comment]
@ruifoto Theoretically speaking, this receiver should work, but I cannot guarantee, of course. And you will get spare mouse :)
https://www.ebay.com/c/23011031758?iid=333381972815

**Comment by doctor64** [comment]
@akumria Please, can you test PR using @pfpfs instruction?

**Comment by pfps** [comment]
@ruifoto  I think you are out of luck - only certain Logitech devices pair with Unifying receivers.   If your mouse and keyboard don't have the Unifying logo on them they may not pair with Unifying receivers.  For example, some K270 keyboards and M185 mice will only pair with Nano receivers with USB id 046d:c534.

**Comment by ruifoto** [comment]
@pfps @doctor64 Thank you both. I will just buy a new set of keyboard then.

**Comment by billyjoker** [comment]
Anoyone knows if EX100 receiver is compatible with the mouse provided with EX110 ?
